### PR TITLE
[nrfconnect] Added EnableKey parameter to factory data

### PR DESCRIPTION
**Pull request by ArekBalysNordic**
#### Problem
Certification tests may make use of the TestEventTrigger command
of GeneralDiagnostics cluster to trigger certain actions.
The command uses authentication with the EnableKey parameter, that
should be placed in the factory data set.

#### Change overview
- Added EnableKey to the JSON Schema.
- Added EnableKey to the factory data.
- Added EnableKey to the Cmake script.
- Added GetEnableKey to the FactoryDataProvider and FactoryDataParser.
- Added new checks to JSON Schema.

#### Testing
Tested by building example with factory data & using nRF52DK.

### CODE REVIEW

File: scripts/tools/nrfconnect/nrfconnect_factory_data.schema [hunk 1]
```diff
@@ -109,12 +113,14 @@
...
             "description": "A key-derivation function for the Symmetric Password-Authenticated Key Exchange.",
             "type": "string",
+            "pattern:": "^hex:{1}",
             "minLength": 36,
             "maxLength": 68
         },
         "spake2_verifier": {
             "description": "A verifier for the Symmetric Password-Authenticated Key Exchange",
             "type": "string",
+            "pattern:": "^hex:{1}",
```
Comment (Damian-Nordic) [inline scripts/tools/nrfconnect/nrfconnect_factory_data.schema:123]: Is that correct? Isn't the pattern supposed to match the entire value, which would require adding some `.*` at the end?

File: scripts/tools/nrfconnect/nrfconnect_factory_data.schema [hunk 2]
```diff
@@ -123,6 +129,13 @@
             "minimum": 0,
             "maximum": 4095
         },
+        "enable_key": {
+            "description": "The Enable Key is a 128-bit value that triggers manufacturer-specific action while invoking the TestEventTrigger Command",
+            "type": "string",
+            "pattern": "^hex:{1}([0-9]|[A,B,C,D,E,F]){32}",
```
Comment (Damian-Nordic) [inline scripts/tools/nrfconnect/nrfconnect_factory_data.schema:135]: Woudn't sth like that work?
```suggestion
            "pattern": "^hex:[0-9A-F]{32}",
```

File: config/nrfconnect/chip-module/Kconfig [hunk 3]
```diff
@@ -254,4 +254,14 @@ config CHIP_DEVICE_ROTATING_DEVICE_UID
 	  A device rotating id unique id which will be generated if
 	  this config is not set in prj.conf file.
 
+config CHIP_DEVICE_ENABLE_KEY
+	string "A 128-bit value that triggers manufacturer-specific action."
```
Comment (Damian-Nordic) [inline config/nrfconnect/chip-module/Kconfig:258]:
```suggestion
	string "Enable Key for triggering test actions on device"
```
Actions are not necessarily manufacturer-specific. CSA can and will define them, too. Also, in general we don't need to use articles ("A something") in Kconfig prompts. Better to keep them shorter :)

File: src/include/platform/DeviceInstanceInfoProvider.h [hunk 4]
```diff
@@ -142,6 +142,18 @@ class DeviceInstanceInfoProvider
      *          if access fails.
      */
     virtual CHIP_ERROR GetRotatingDeviceIdUniqueId(MutableByteSpan & uniqueIdSpan) = 0;
+
```
Comment (Damian-Nordic) [inline src/include/platform/DeviceInstanceInfoProvider.h:145]: So far there's no common code that would use this value, and I think the idea was to let applications initialize handlers of TestEventTriggers, so we may simply implement it in our class and not add it to the common interface.

**Review by Damian-Nordic — APPROVED** [review]
I think we should get rid of "manufacturer-specific" bit from the help texts + I have a few more comments, but LGMT overall.

File: scripts/tools/nrfconnect/nrfconnect_factory_data.schema (hunk 1 above)
Comment (ArekBalysNordic) [reply scripts/tools/nrfconnect/nrfconnect_factory_data.schema:123]: I guess that's correct, according to https://json-schema.org/understanding-json-schema/reference/regular_expressions.html#regular-expressions . `^` means the specific regular expression must be placed at the beginning of the string, and `{1}` means at most only one occurrence.

File: examples/all-clusters-app/nrfconnect/main/AppTask.cpp [hunk 5]
```diff
@@ -59,10 +59,10 @@ K_MSGQ_DEFINE(sAppEventQueue, sizeof(AppEvent), APP_EVENT_QUEUE_SIZE, alignof(Ap
 
 namespace {
 
-// NOTE! This key is for test/certification only and should not be available in production devices.
-// Ideally, it should be a part of the factory data set.
-constexpr uint8_t kTestEventTriggerEnableKey[16] = { 0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77,
-                                                     0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff };
+// NOTE! This key is for test/certification only and should not be available in production devices!
+// To use production-ready Enable Key add it to the factory data and enable CONFIG_CHIP_FACTORY_DATA config.
```
Comment (Damian-Nordic) [inline examples/all-clusters-app/nrfconnect/main/AppTask.cpp:63]: "production-ready" can be confusing given that we've just said that it's not supposed to be used in production. Maybe: "if `CONFIG_CHIP_FACTORY_DATA` is enabled, this value is read from the factory data". Also, nit: `static` is redundant if this is in anonymous namespace.

File: examples/all-clusters-app/nrfconnect/main/AppTask.cpp [hunk 6]
```diff
@@ -191,12 +191,16 @@ CHIP_ERROR AppTask::Init()
     SetDeviceInstanceInfoProvider(&mFactoryDataProvider);
     SetDeviceAttestationCredentialsProvider(&mFactoryDataProvider);
     SetCommissionableDataProvider(&mFactoryDataProvider);
+    // Read EnableKey from the factory data.
+    MutableByteSpan enableKey(sTestEventTriggerEnableKey);
+    err = mFactoryDataProvider.GetEnableKey(enableKey);
```
Comment (Damian-Nordic) [inline examples/all-clusters-app/nrfconnect/main/AppTask.cpp:196]: I think we shouldn't initialize the delegate if this fails. Alternatively, initialize `sTestEventTriggerEnableKey` to all zeros if `CONFIG_CHIP_FACTORY_DATA` and ignore an error here (assignment to `err` is not needed if we don't check the value).

File: examples/light-switch-app/nrfconnect/main/AppTask.cpp [hunk 7]
```diff
@@ -65,6 +66,11 @@ K_MSGQ_DEFINE(sAppEventQueue, sizeof(AppEvent), kAppEventQueueSize, alignof(AppE
 Identify sIdentify = { kLightEndpointId, AppTask::IdentifyStartHandler, AppTask::IdentifyStopHandler,
                        EMBER_ZCL_IDENTIFY_IDENTIFY_TYPE_VISIBLE_LED };
 
+// NOTE! This key is for test/certification only and should not be available in production devices.
+// Ideally, it should be a part of the factory data set.
```
Comment (Damian-Nordic) [inline examples/light-switch-app/nrfconnect/main/AppTask.cpp:70]: Comment should be updated if you changed it in other examples.

File: scripts/tools/nrfconnect/generate_nrfconnect_chip_factory_data.py [hunk 8]
```diff
@@ -320,13 +322,16 @@ def allow_any_int(i): return int(i, 0)
     optional_arguments.add_argument("--pai_cert", type=str,
                                     help="[.der] Provide the path to .der file containing PAI certificate.")
     optional_arguments.add_argument("--rd_uid", type=str,
-                                    help="[hex string] Provide the rotating device unique ID. If this argument is not provided a new rotating device id unique id will be generated.")
+                                    help="[hex string] [pattern: 'FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF'] Provide the rotating device unique ID. If this argument is not provided a new rotating device id unique id will be generated.")
```
Comment (Damian-Nordic) [inline scripts/tools/nrfconnect/generate_nrfconnect_chip_factory_data.py:325]: Not sure if this pattern thing is understandable. IMO, 128-bit **hex-encoded** value is good enough.

**Review by Damian-Nordic — APPROVED** [review]

File: scripts/tools/nrfconnect/nrfconnect_factory_data.schema (hunk 1 above)
Comment (ArekBalysNordic) [reply scripts/tools/nrfconnect/nrfconnect_factory_data.schema:126]: Oh, I see now what you meant. I agree the following part should contain only 0-9, A-F, and a-f symbols. I will add this one: ^hex:{1}([0-9A-Fa-f]){1,} and now after the "hex:" prefix there will be only a proper string that contains specific symbols.

Comment (Damian-Nordic) [reply scripts/tools/nrfconnect/nrfconnect_factory_data.schema:126]: Right, also `{1}` is rather redundant :)

File: examples/all-clusters-app/nrfconnect/main/AppTask.cpp (hunk 6 above)
Comment (ArekBalysNordic) [reply examples/all-clusters-app/nrfconnect/main/AppTask.cpp:196]: I just forgot to add checking here :/

**Pull request closed without merging**